doc/persistence.adoc

```diff
@@ -0,0 +1,165 @@
+* Size of the data, frequency of changes to the data: Can it fit into memory? How expensive can updates be?
+* Granularity needed: How often must the data be persisted?
+* Required queryability of the data: Are indices for fast lookups required?
+* Required data access, objects the data is tied to: Items? Players?
+
+== (De-)Serialization
+
+=== JSON
+
+Minetest provides a JSON serializer based on `jsoncpp`.
+
+The following types are supported by JSON:
+
+* booleans
+* numbers
+* strings
+* tables (objects/arrays)
+
+NOTE: Negative and positive infinity are represented by numbers which exceed the double number range by a lot (plus/minus `1e9999`).
+
+WARNING: The empty table `{}` (ambiguous: might be either `{}` or `[]`) and `nan` aren't supported by JSON and will be turned into `null`.
+
+Tables may either:
+. Contain only positive integer keys (represented as array) or
+. Contain only string keys (represented as object/dictionary)
+obviously, plenty of Lua tables aren't representable this way (boolean keys, mixed hash/list keys, fractional keys, inf keys)
+
+==== `minetest.write_json(data, [styled])`
+
+If the data is serializable as outlined above:: Returns a JSON string representing `data`. Formats the JSON nicely to improve readability if `styled` is truthy.
+Else:: Returns `nil` and one of the following errors if it fails:
+* `Can't use indexes with fractional part in JSON`
+* `Lua key to convert to JSON is not a string or number`
+* `Can't mix array and object values in JSON`
+
+WARNING: Hash tables containing only positive integer keys - a table of `minetest.hash_node_position` hashes for instance - will be turned into *an array with holes*, that is, all `nil` values from `1` to the maximum index will be filled with `null`, which means terrible performance and possibly very large output generated from very small input. *Do not allow direct access to `minetest.write_json`* as it can be trivially used to DoS your server.
+
+TIP: Sparse hash maps, should always be converted to objects, using only string keys, for JSON. You might do this just for serialization.
+
+WARNING: The JSON serializer is furthermore limited by its recursion depth: Only a recursion depth up to *16* is allowed. Very nested data structures can't be (de)serialized. Circular references will cause the JSON serializer to recurse until the maximum recursion depth is exceeded.
+
+TIP: Use the `json = assert(minetest.write_json(data))` idiom to not silently ignore errors when serializing.
+
+==== `minetest.parse_json(json, [nullvalue])`
+
+If `json` is valid JSON:: Returns the Lua value represented by the JSON string `json`, with `null` values deserialized to `nullvalue` (which defaults to `nil`).
+Else:: Returns `nil` and logs an error.
+
+TIP: If you must use JSON (to interact with a web interface, for instance), consider using your JSON library of choice; Lua-only implementations are available.
```

Comment:
Can you give an example of one which solves the issues presented by the Minetest implementation?

Reply:
I have implemented my own and didn't want to mention that. https://github.com/grafi-tt/lunajson/ could be mentioned here.